### PR TITLE
(MODULES-3034) Chocolatey Feature support

### PR DESCRIPTION
**Pull request by jpogran**
Support enabling or disabling chocolatey features

### CODE REVIEW

File: spec/unit/puppet/provider/chocolateyfeature/windows_spec.rb [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    <feature name="usePackageExitCodes" enabled="true" setExplicitly="false" description="Use Package Exit Codes - Package scripts can provide exit codes. With this on, package exit codes will be what choco uses for exit when non-zero (this value can come from a dependency package). Chocolatey defines valid exit codes as 0, 1605, 1614, 1641, 3010. With this feature off, choco will exit with a 0 or a 1 (matching previous behavior). Available in 0.9.10+." />
+  </features>
+  <apiKeys>
+    <apiKeys source="https:\/\/chocolatey.org\/" key="bogus\/encrypted+value=" />
+ </apiKeys>
+</chocolatey>
+  EOT
+  }
+
+  let (:newer_choco_version) {'0.9.10.0'}
```
Comment (glennsarti) [inline spec/unit/puppet/provider/chocolateyfeature/windows_spec.rb]: This does not seem to be used

File: spec/unit/puppet/provider/chocolateyfeature/windows_spec.rb [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  let (:minimum_supported_version) {'0.9.9.0'}
+  let (:last_unsupported_version) {'0.9.8.33'}
```
Comment (glennsarti) [inline spec/unit/puppet/provider/chocolateyfeature/windows_spec.rb]: This does not seem to be used

File: lib/puppet/provider/chocolateyfeature/windows.rb [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    feature[:ensure] = :disabled
+    feature[:ensure] = :enabled if enabled
+
+    Puppet.debug("Loaded feature '#{feature.inspect}'.")
+
+    feature
+  end
+
+  def self.features
+    @features ||=  get_features.collect do |item|
```
Comment (glennsarti) [inline lib/puppet/provider/chocolateyfeature/windows.rb]: I'm not 100% sure why we're using a class level variable for features here.  The only other method that uses `@features` is `self.refresh_features` and that's only called by flush.

Should we scrap using the class level features variable and just query the filesystem every time?

File: lib/puppet/provider/chocolateyfeature/windows.rb [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    args << '--name' << resource[:name]
+
+    begin
+      Puppet::Util::Execution.execute([command(:chocolatey), *args])
+    rescue Puppet::ExecutionFailure
+      raise Puppet::Error, "An error occurred running choco. Unable to set Chocolatey source configuration for #{self.inspect}"
+    end
+
+    @property_hash.clear
+    @property_hash = { :ensure => ( @property_flush[:ensure] )}  #if  @property_flush[:ensure] == :absent
```
Comment (glennsarti) [inline lib/puppet/provider/chocolateyfeature/windows.rb]: Is this comment still required?

File: spec/unit/puppet/provider/chocolateyfeature/windows_spec.rb [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+require 'puppet/provider/chocolateyfeature/windows'
+require 'rexml/document'
+
+provider = Puppet::Type.type(:chocolateyfeature).provider(:windows)
+describe provider do
+  let (:name) { 'allowglobalconfirmation' }
+  let (:resource) { Puppet::Type.type(:chocolateyfeature).new(:provider => :windows, :name => name) }
+  let (:choco_config) { 'c:\choco.config' }
+  let (:choco_config_contents) { <<-'EOT'
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (glennsarti) [inline spec/unit/puppet/provider/chocolateyfeature/windows_spec.rb]: This config file seems to be replication in the sources tests.  Should this become a test fixture instead?

Comment (jpogran) [inline spec/unit/puppet/provider/chocolateyfeature/windows_spec.rb]: Both operate on the same xml config file, but look at different parts of it. I largely kept it the same, but can see the need for having differences in the future.

File: lib/puppet/provider/chocolateyfeature/windows.rb (hunk 3 above)
Comment (jpogran) [inline lib/puppet/provider/chocolateyfeature/windows.rb]: This was carried over from sources, so we'll have to ask @ferventcoder there

File: spec/unit/puppet/provider/chocolateyfeature/windows_spec.rb (hunk 5 above)
Comment (glennsarti) [inline spec/unit/puppet/provider/chocolateyfeature/windows_spec.rb]: Ok...can leave it as is then

File: lib/puppet/provider/chocolateyfeature/windows.rb [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+require 'rexml/document'
+
+Puppet::Type.type(:chocolateyfeature).provide(:windows) do
+  confine :operatingsystem => :windows
+  defaultfor :operatingsystem => :windows
+
+  require Pathname.new(__FILE__).dirname + '../../../' + 'puppet_x/chocolatey/chocolatey_common'
+  include PuppetX::Chocolatey::ChocolateyCommon
+
+  MINIMUM_SUPPORTED_CHOCO_VERSION = '0.9.9.0'
```
Comment (Iristyle) [inline lib/puppet/provider/chocolateyfeature/windows.rb]: I think this constant definition might clash with the one of the same name in the sources code.

Comment (glennsarti) [inline lib/puppet/provider/chocolateyfeature/windows.rb]: Is that constant scoped to the class or global?

Comment (jpogran) [inline lib/puppet/provider/chocolateyfeature/windows.rb]: The test output goes by so fast, you don't see that we do get a warning for duplicate const.

The problem here we are trying to solve is that different capabilities were introduced in different minor versions and we need to have a bare minimum before we can even run. `source` happens to have the same version as `feature` coincidently, but `source` also tracks more versions. There are many point versions where features (ever use a word too many times and now it loses meaning?) were introduced, so I can easily see us adding more as we add other types. 

I would suggest making a single constant, but it's just coincidence that these two share the same version and we would be back in the same boat. I agree this should be named, but should we also change the other test to align with this?

File: lib/puppet/provider/chocolateyfeature/windows.rb (hunk 3 above)
Comment (jpogran) [inline lib/puppet/provider/chocolateyfeature/windows.rb]: After reviewing this, I thought there may be more penalty accessing disk and reading the XML. I tested time for both it there and it removed. There are no perceptible differences in time, and it adds another place to question. Removed the global variables.

**Comment by ferventcoder** [comment]
Such a long commit body you have there 👍

**Comment by ferventcoder** [comment]
This looks great, doing final review and then I will pull this in. To help out I will also create the acceptance specs.

**Comment by ferventcoder** [comment]
Superseded by https://github.com/puppetlabs/puppetlabs-chocolatey/pull/24

**Pull request closed without merging**